### ISSUE TITLE
Hi from upstream

### ISSUE THREAD

**Issue by rurban** [benegon/libredwg#1]
Hi David,
Did you file a copyright disclaimer or waiver with the FSF?
I'm just looking at your patches, and they are looking good. Nothing earth shattering, but useful.

I've added a mirror at https://github.com/LibreDWG/libredwg, where you could add PR's easily.

* const'ing the API is trivial. I've also added some const'ing in the meantime. I could do that by myself, if you haven't signed.

* The hashmap is to speed up performance of the handle resolver. LibreDWG's concern is not performance, more correctness and ease of use, but why not. Did you measure some impact with big DWG's?
I've also looked at sorting the refs, as already done in the encoder, and then use binary search. I'm not sure yet which approach to use. We already have the sorting routine, I'm just not using it with the decoder, as it will change all refs.
And for the hashmap we would need a copyright assigned hashmap.

My merge is here: https://github.com/LibreDWG/libredwg/tree/benegon

**Comment by codehero** [comment]
Hi rurban, so glad to see an active maintainer for this project.
It's been awhile since I've touched this code, but I used some very large files and the hashmap helped as I remember. I wasn't particularly concerned about copyright so I just used Android.

Feel free to use any of my changes.
How do I do the waiver?

**Comment by rurban** [comment]
Now I got a 40MB dwg which is hugely slow, where using your hash table brought it down from 2m to 5s. I see now your point :)

The waiver url is described in the CONTRIBUTING document. But since this is an Android library, we would need a waiver from their authors.

So I just went ahead and wrote my own mast intmap in the last hour from scratch.
I didn't like the overblown Android library with seperate chaining anyway. A hash table should be simple and fast, use open addressing and we don't need deletion, and can avoid key 0 and value 0.

Thanks anyway for proving the point that we need a hash for realistic DWGs. Very appreciated.
